server/src/main/java/org/opensearch/gateway/remote/model/RemoteIndexMetadata.java

```diff
@@ -0,0 +1,94 @@
+import org.opensearch.gateway.remote.RemoteClusterStateUtils;
+import org.opensearch.index.remote.RemoteStoreUtils;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+import org.opensearch.repositories.blobstore.ChecksumBlobStoreFormat;
+
+/**
+ * Wrapper class for uploading/downloading {@link IndexMetadata} to/from remote blob store
+ */
+public class RemoteIndexMetadata extends AbstractRemoteBlobObject<IndexMetadata> {
+
+    public static final int INDEX_METADATA_CURRENT_CODEC_VERSION = 1;
+
+    public static final ChecksumBlobStoreFormat<IndexMetadata> INDEX_METADATA_FORMAT = new ChecksumBlobStoreFormat<>(
+        "index-metadata",
+        METADATA_NAME_PLAIN_FORMAT,
+        IndexMetadata::fromXContent
+    );
+    public static final String INDEX_PATH_TOKEN = "index";
+
+    private IndexMetadata indexMetadata;
+
+    public RemoteIndexMetadata(IndexMetadata indexMetadata, String clusterUUID, BlobStoreRepository blobStoreRepository) {
+        super(blobStoreRepository, clusterUUID);
+        this.indexMetadata = indexMetadata;
+    }
+
+    public RemoteIndexMetadata(String blobName, String clusterUUID, BlobStoreRepository blobStoreRepository) {
+        super(blobStoreRepository, clusterUUID);
+        this.blobName = blobName;
+    }
+
+    @Override
+    public IndexMetadata get() {
+        return indexMetadata;
+    }
+
+    @Override
+    public BlobPathParameters getBlobPathParameters() {
+        return new BlobPathParameters(List.of(INDEX_PATH_TOKEN, indexMetadata.getIndexUUID()), "metadata");
+    }
+
+    @Override
+    public String generateBlobFileName() {
+        String blobFileName = String.join(
+            RemoteClusterStateUtils.DELIMITER,
+            getBlobPathParameters().getFilePrefix(),
+            RemoteStoreUtils.invertLong(indexMetadata.getVersion()),
+            RemoteStoreUtils.invertLong(System.currentTimeMillis()),
+            String.valueOf(INDEX_METADATA_CURRENT_CODEC_VERSION) // Keep the codec version at last place only, during reads we read last
+            // place to determine codec version.
+        );
+        this.blobFileName = blobFileName;
+        return blobFileName;
+    }
+
+    @Override
+    public UploadedMetadata getUploadedMetadata() {
+        assert blobName != null;
+        return new UploadedIndexMetadata(indexMetadata.getIndexName(), indexMetadata.getIndexUUID(), blobName);
+    }
+
+    @Override
+    public InputStream serialize() throws IOException {
+        return INDEX_METADATA_FORMAT.serialize(indexMetadata, generateBlobFileName(), getCompressor(), RemoteClusterStateUtils.FORMAT_PARAMS)
+            .streamInput();
+    }
+
+    @Override
+    public IndexMetadata deserialize(InputStream inputStream) throws IOException {
+        // Blob name parameter is redundant
+        return INDEX_METADATA_FORMAT.deserialize(blobName, getBlobStoreRepository().getNamedXContentRegistry(), Streams.readFully(inputStream));
```

Comment:
is BlobStoreRepository added to AbstractRemoteBlobObject just for NamedXContextRegistry() you can wire that directly instead.

Reply:
It is added to get the compressor as well for serialization in the `serialize` method above. Are you suggesting to wire both compressor and NamedXContextRegistry directly ?